presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriterFactory.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.hadoop.hive.serde2.columnar.ColumnarSerDe;
+import org.apache.hadoop.hive.serde2.columnar.LazyBinaryColumnarSerDe;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
+import org.joda.time.DateTimeZone;
+
+import javax.inject.Inject;
+
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Optional;
+import java.util.Properties;
+
+import static com.facebook.presto.hive.HiveType.toHiveTypes;
+import static com.facebook.presto.hive.rcfile.RcFilePageSourceFactory.createTextVectorEncoding;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+import static org.apache.hadoop.hive.metastore.api.hive_metastoreConstants.META_TABLE_COLUMNS;
+import static org.apache.hadoop.hive.metastore.api.hive_metastoreConstants.META_TABLE_COLUMN_TYPES;
+
+public class RcFileFileWriterFactory
+        implements HiveFileWriterFactory
+{
+    private final DateTimeZone hiveStorageTimeZone;
+    private final HdfsEnvironment hdfsEnvironment;
+    private final TypeManager typeManager;
+    private final NodeVersion nodeVersion;
+
+    @Inject
+    public RcFileFileWriterFactory(
+            HiveClientConfig hiveClientConfig,
+            HdfsEnvironment hdfsEnvironment,
+            TypeManager typeManager,
+            NodeVersion nodeVersion)
+    {
+        requireNonNull(hiveClientConfig, "hiveClientConfig is null");
+        this.hiveStorageTimeZone = hiveClientConfig.getDateTimeZone();
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.nodeVersion = requireNonNull(nodeVersion, "nodeVersion is null");
+    }
+
+    @Override
+    public Optional<HiveFileWriter> createFileWriter(
+            Path path,
+            List<String> inputColumnNames,
+            StorageFormat storageFormat,
+            Properties schema,
+            JobConf configuration,
+            ConnectorSession session)
+    {
+        if (!HiveSessionProperties.isRcfileOptimizedWriterEnabled(session)) {
+            return Optional.empty();
+        }
+
+        if (!RCFileOutputFormat.class.getName().equals(storageFormat.getOutputFormat())) {
+            return Optional.empty();
+        }
+
+        RcFileEncoding rcFileEncoding;
+        if (LazyBinaryColumnarSerDe.class.getName().equals(storageFormat.getSerDe())) {
+            rcFileEncoding = new BinaryRcFileEncoding();
+        }
+        else if (ColumnarSerDe.class.getName().equals(storageFormat.getSerDe())) {
+            rcFileEncoding = createTextVectorEncoding(schema, hiveStorageTimeZone);
+        }
+        else {
+            return Optional.empty();
+        }
+
+        Optional<String> codecName = Optional.ofNullable(configuration.get(FileOutputFormat.COMPRESS_CODEC));
+
+        // existing tables may have columns in a different order
```

Comment:
I edited the original comment to fix some typo.

I believe I was correct after I went through the relevant code again.